src/generate_features.py

```diff
@@ -1,3 +1,215 @@
+        cov_type = config.get('cov_type')
+        labels = gmm_clust(X, n_components, cov_type)
+    elif config.get('cluster-method') == 'kmeans':
+        k = config.get('k')
+        labels = km_clust(X, k)
+
+    logger.debug('Cluster solution score: %f', score_ch(X, labels))
+    logger.info('Cluster dispersion: %s',
+                np.unique(labels, return_counts=True))
+
+    order_types['label'] = labels
+
+    if config.get('save_cluster_heatmap'):
+        gen_plots(order_types)
+
+    # Limit orders to "prior" set for training data
+    logger.debug('Engineering shopper features')
+    orders = orders[orders['eval_set'] == 'prior']
+    users = pd.DataFrame(orders.groupby('user_id').size().rename('n_orders'))
+
+    # Get num orders per day of week for each user
+    dow_counts = orders.pivot_table(index='user_id', columns='order_dow',
+                                    values='order_number', aggfunc='count')
+    dow_counts.columns = ['n_dow_{}'.format(col)
+                          for col in dow_counts.columns]
+    users = users.join(dow_counts.fillna(0))
+
+    # Get num orders per hour of day for each user
+    hod_counts = orders.pivot_table(index='user_id', columns='order_hour_of_day',
+                                    values='order_number', aggfunc='count')
+    hod_counts.columns = ['n_hod_{}'.format(col) for col in hod_counts.columns]
+    users = users.join(hod_counts.fillna(0))
+
+    # Capture each user's typical order size with summary statistics
+    order_size_stats = orders.groupby('user_id').agg({
+        'order_size': [np.mean, np.std, np.max, np.min]
+        })
+    # Flatten hierarchical column names
+    order_size_stats.columns = ['_'.join(col).strip()
+                                for col in order_size_stats.columns.values]
+
+    users = users.join(order_size_stats)
+    size_count = orders.pivot_table(index='user_id', columns='size_cat',
+                                    values='order_number', aggfunc='count')
+    users = users.join(size_count.fillna(0))
+
+    # Describe users by the composition of their typical order
+    groups = full.groupby('user_id')
+    cols = ['days_since_prior_order', 'reordered', 'organic', 'popular']
+    cols += cats
+    means = groups[cols].mean()
+    users = users.join(means)
+
+    # Add target variable (order_type for next purchase)
+    train = order_types[order_types['eval_set'] == 'train']
+    user_targets = order_types[order_types['eval_set'] == 'train']
+    user_targets = user_targets.groupby('user_id')['label'].first()
+    users = users.join(user_targets.rename('label'))
+    users = users[~users.label.isna()]
+
+    shoppers_path = os.path.join(ROOT, 'data', 'features', 'shoppers.csv')
+    orders_path = os.path.join(ROOT, 'data', 'features', 'baskets.csv')
+    logger.info('Feature engineering complete. Saving output...\n\t%s\n\t%s',
+                shoppers_path, orders_path)
+    users.to_csv(shoppers_path)
```

Comment:
would it be useful to save a local in auxiliary of something for use in later model building